### ISSUE TITLE
(Bug) Threading issue in CacheContext

### ISSUE THREAD

**Issue by vvdb-architecture** [GFlisch/Arc4u.Guidance.Doc#45]
The [CacheContext](https://github.com/GFlisch/Arc4u/blob/master/src/Arc4u.Standard.Caching/CacheContext.cs) is a shared instance. 
This means that multiple threads can access its public members:

~~~csharp
  public ICache Default { get; }
  public bool Exist(string cacheName);
  public ICache this[string cacheName]
~~~

Of these 3 public members, only the last one, ` public ICache this[string cacheName]`, changes the state of the instance. The `Default` property indirectly calls this property as well. The code looks like this now:

~~~csharp
 public ICache this[string cacheName]
        {
            get
            {
                // look inside the collection if the cache is initialized.
                if (_caches.ContainsKey(cacheName))
                    return _caches[cacheName];

                // look if the cache is not yet started and start it.
                lock (_lock)
                {
                            (the code here eventually changes _caches and _uninitializedCaches)
                }

                throw new InvalidOperationException($"There is no cache configured with the name {cacheName}.");
            }

        }
~~~

The code is wrong because the `_lock` only protects the code changing the state of `_caches` and `uninitializedCaches`, but does not protect them against threads reading their state at the same time.
It is therefore possible that one thread changes `_caches`/`uninitializedCaches` while another thread reads from `_caches`/`uninitializedCaches` at the same time. Since both are of type `Dictionary<TKey,TValue>` and that type is not thread-safe, the behavior is undefined.

For a better explanation, see https://stackoverflow.com/questions/710696/containskey-thread-safe or https://stackoverflow.com/questions/73438935/is-it-thread-safe-to-lock-a-net-dictionary-on-write-and-read-it-lock-free.

This issue has been found while investigating #44 

This needs to be rewritten to use either by extending the locking, a Reader/Writer lock, a pair of ConcurrentDictionary types (but be careful) or the atomic exchange pattern.

**Comment by devbrsa** [comment]
`ReadWriteLockSlim` seems a good option here, in this bug context.
https://learn.microsoft.com/en-us/dotnet/api/system.threading.readerwriterlockslim?view=net-6.0

**Comment by vvdb-architecture** [comment]
If you want a resource to be accessible for both readers and writers,  `ReaderWriterLockSlim` is the optimal choice if the following conditions are simultaneously true:
- the readers outnumber the writers
- the ratio between the number of readers and the number of writers doesn't vary much

The reason for the second condition is that `ReaderWriterLockSlim` introduces overhead for both readers and writers, since you need to acquire locks in both cases (even if the lock types are different).

However, look at behavior of `CacheContext`.  This is a class holding a very small number of instances, which are initialized on demand. It's that demand-based initialization that is our single "writer" . Moreover, for a specific instance, this writing happens only once. Therefore, in the lifetime of the application, it is reasonable to assume that over time, the number of "writers" will be zero (because all caches that are needed will have been initialized), and we'll have only readers.

With this behavior, it seems wasteful to have readers incur the overhead of a lock, even if it's a read lock.

I would propose to use an atomic exchange patter instead. Only the accessor needs to be changed:

~~~csharp
        public ICache this[string cacheName]
        {
            get
            {
                // look inside the collection if the cache is initialized.
                if (_caches.ContainsKey(cacheName))
                    return _caches[cacheName];

                // look if the cache is not yet started and start it.
                lock (_lock)
                {
                    // double-check for initialization status since it might have changed
                    if (_caches.ContainsKey(cacheName))
                        return _caches[cacheName];

                    if (_uninitializedCaches.ContainsKey(cacheName))
                    {
                        try
                        {
                            var cache = _dependency.Resolve<ICache>(_uninitializedCaches[cacheName]);

                            cache.Initialize(cacheName);

                            var caches = new Dictionary<string, ICache>(_caches);
                            var uninitializedCaches = new Dictionary<string, string>(_uninitializedCaches);
                            caches.Add(cacheName, cache);
                            uninitializedCaches.Remove(cacheName);

                            // atomic exchange of object state. Note that the order matters for efficiency
                            Interlocked.Exchange(ref _caches, caches);
                            Interlocked.Exchange(ref _uninitializedCaches, uninitializedCaches);
                            return cache;
                        }
                        catch (Exception ex)
                        {
                            _logger.Technical().From<CacheContext>().Exception(ex).Log();
                        }
                    }
                }

                throw new InvalidOperationException($"There is no cache configured with the name {cacheName}.");
            }

        }
~~~


In any case, `CacheContect` is a very strange class: it does a named lookup of instances that are initialized on first use. This is exactly what named-based dependency injection does.  There is the notion of "AutoStart", but we should investigate if this can't be refactored somehow into the existing DI. This is the subject of another issue.
